### ISSUE TITLE
Uncontrolled state spotted on joints controlled via MC4Plus

### ISSUE THREAD

**Issue by S-Dafarra** [robotology/icub-tech-support#673]
### Description of the failure
The left arm goes in a state which is not down, i.e. the motors are not completely switched off, but yet the joints starting from the wrist prono-supination cannot be controlled and are kind of "compliant".

### Detailed conditions and logs of the failure
[logArmDown_part1.txt](https://github.com/robotology/icub-support/files/2102821/logArmDown_part1.txt)
[logArmDown_part2.txt](https://github.com/robotology/icub-support/files/2102823/logArmDown_part2.txt)

Image of the arm:
![image](https://user-images.githubusercontent.com/18591940/41419980-a584a70e-6ff3-11e8-82d0-8cc40fe074eb.png)
I can move easily the prono-supination, while the wrist is stuck to the upper limit.

cc @julijenv

**Comment by S-Dafarra** [comment]
Also there is a super annoying whistle coming from the hand. If we squeeze the fingers (moving the ``l_hand_fingers`` joint), it stops.

**Comment by julijenv** [comment]
Hi @S-Dafarra , still on going this problem?!?!

**Comment by S-Dafarra** [comment]
Yes, actually it happens on both arms. For example, yesterday it happened on the right arm. Interestingly we also pressed the fault, but the forearm remained in that strange state.

**Comment by S-Dafarra** [comment]
This issue is still occurring.

**Comment by S-Dafarra** [comment]
It just happened today, again on the left arm. As far as I know, this is still happening on **both** arms.
Here a more recent log 
[log_icub-head_unresponsiveWrist.txt](https://github.com/robotology/icub-tech-support/files/3690179/log_icub-head_unresponsiveWrist.txt)

After a quick F2F talk with @julijenv we suspect this issue may be related to a board resetting, maybe after a voltage drop, which causes the board to remain in an uncontrolled state with a constant PWM output. 
Maybe @valegagge or @marcoaccame may have some opinion on this.

**Comment by julijenv** [comment]
Hi @valegagge ,
if you time to check that and give me your thoughts about it I'd be happy,
thx in advacnce

**Comment by S-Dafarra** [comment]
The issue is still occurring.

**Comment by S-Dafarra** [comment]
It still happens.

**Comment by S-Dafarra** [comment]
After a quick f2f with @julijenv we understood that we need to check whether the board is still pingable when this problem is occurring. This may give us some more hints of what is going on.

**Comment by S-Dafarra** [comment]
I am not going to give up on this ``stale-bot``! :joy:

**Comment by S-Dafarra** [comment]
Yet another log. BY accident in the end we switched off all the boards by mistake before stopping the ``yarprobotinterface``. Nevertheless, the left wrist was in this status since a while.

[log_icub-head_yarprobotinterface_1705_left_wrist_uncontrollable.txt](https://github.com/robotology/icub-tech-support/files/4938892/log_icub-head_yarprobotinterface_1705_left_wrist_uncontrollable.txt)

**Comment by S-Dafarra** [comment]
Still happening

**Comment by pattacini** [comment]
Let's try to tackle this with the help of @davidetome

**Comment by davidetome** [comment]
Ciao,

I took a look at the logs provided. The fault that causes this behaviour is most likely : 

```
 6450 1461.226705 <ERROR>   FATAL: EthResource::getRemoteValue() DID NOT have replies from BOARD left_arm-eb24-j4_7 with IP 10.0.1.24  even after 0.200232 seconds: CANNOT PROCEED ANY FURTHER
```

As already pointed out in this [comment](https://github.com/robotology/icub-tech-support/issues/673#issuecomment-538350914), does it seem that the board is resetting for some reasons (a short circuit in a certain position of the arm? a wire not well connected?  A fault on the `mc4plus` board?)

I propose to do some tests tomorrow morning but I need help, maybe we need to uncover the arm, check for connections and maybe replacing the board in the last instance. @julijenv are you available for this?  

@S-Dafarra just a couple of questions:

1. Is the robot available tomorrow morning?
2. Is the failure replicable in some conditions (positions) or it happens randomly?
3. Are the `firmware` of the `boards` updated? If not, can we do that? 
4. Is the sofware running on `laptop/server` and on the `icub-head`updated?

I look forward to your replies,  thanks!

cc @pattacini

**Comment by pattacini** [comment]
They're doing many tests these days with Greeny. Let's see if the robot is available.

**Comment by S-Dafarra** [comment]
Hi @davidetome 
> * Is the robot available tomorrow morning?

Probably yes. We are planning to use it today full day, but tomorrow it should be free. The robot is still reserved for the iRonCub experiments, but they got postponed of one week. Unfortunately, we weren't able to cancel the booking. I think tomorrow morning is ok.



> * Is the failure replicable in some conditions (positions) or it happens randomly?

If it is repeatable, we did not find how. It happens from time to time, especially under stress (when walking or doing other dynamic demos), even without moving the arms. Once, during a demo, it was happening by simply waving the hand. In general, it seems to happen only when moving. Unfortunately, it seems pretty random.

> * Are the `firmware` of the `boards` updated? If not, can we do that?

I tried to update about two months ago, but if you need to update the boards, feel free to do it. If problems occur, we would need to fix them anyway :wink:  

> 4\. Is the sofware running on `laptop/server` and on the `icub-head`updated?

Right now we are using the latest release on the ``icub-head``. The configuration files are over ``master``. 
Anyways, I remember this issue happening since the first time we started the robot after being assembled.

Even if we are not spotting the actual issue, do you think it may be possible to work it around via sfotware?

**Comment by davidetome** [comment]
Ciao @S-Dafarra ,

 Answering your question I think it is not so easy to work it around also because it involves both `FW` and `SW`, especially if the board is resetting as it seems; I mean when `yarprobotinterface` starts configures all the boards and services and if a board goes down and up again it can't be anymore usable until  `yarprobotinterface` restarts.
But on this point @marcoaccame is certainly the most suitable person.

 I spoke with @julijenv, and he will be at CRIS until about 12.00.
He also noticed that the robot is just felt down so is going to check if there are damages and let me know if the robot is still usable. 
If yes, I can come and start trying to reproduce the failure checking for the `EB24` board reset. 

Anyway, can you please advise your colleagues and leave the robot powered off so tomorrow I can bring it to `proto` lab?

cc @pattacini @Fabrizio69

**Comment by S-Dafarra** [comment]
> Answering your question I think it is not so easy to work it around also because it involves both `FW` and `SW`, especially if the board is resetting as it seems; I mean when `yarprobotinterface` starts configures all the boards and services and if a board goes down and up again it can't be anymore usable until `yarprobotinterface` restarts.

I see. We usually reboot also the boards. I may be wrong, but restarting the `yarprobotinterface` may not be enough. I was hoping that with some ``yarpmotorgui`` macumba it was possible to get them back (by triggering the calibration again maybe), but I am just wondering. I totally understand the complexity.



> He also noticed that the robot is just felt down

Yeah, I have some responsibility :grin: 



> let me know if the robot is still usable.

For this I would trust the opinion of @julijenv 

> Anyway, can you please advise your colleagues and leave the robot powered off so tomorrow I can bring it to proto lab?

Yes, I will either do it by myself or ask someone to do it.

**Comment by davidetome** [comment]
Juast an update, the robot is not starting due to a problem on the power supply (backpack, some mechanical parts are bended).
@julijenv  isdisassembling the part and going to check

**Comment by davidetome** [comment]
Another Update,

The robot now starts again (it was a problem with `BMON` board), @julijenv straightened the bent mechanical part.
We faced another issue with the `EXT FAULT` that was seen pressed by `yarprobotinterface` tough the button was not really pressed.
Davide Gandini checked the wires and cabling fixing it (probably the connector not well inserted after the fall.

Now I'm going to move the `left-arm` for a while using a sequence in `yarpmotorgui` and checking if the board `EB24` restarts

**Comment by davidetome** [comment]
Las t update

the robot is running this [sequence](https://github.com/robotology/icub-tech-support/files/5277083/la.posleft_arm.txt) on the `left-arm` for about 2 hours and the board `EB24` is still up and running.

I propose to plan another day (at least) with @julijenv help to do some tests uncovering the arm.

@S-Dafarra you can take back the robot for the moment :+1: 

cc @pattacini

**Comment by S-Dafarra** [comment]
> the robot is running this [sequence](https://github.com/robotology/icub-tech-support/files/5277083/la.posleft_arm.txt) on the `left-arm` for about 2 hours and the board `EB24` is still up and running.

Thanks a lot @davidetome. Just to understand, more or less which is the motion of the arm you were trying? Unfortunately, we never understood the root cause. Sometimes it happened also when running the torque-controlled Yoga++, where the arms are moved very little. Also, it happens on both arms.

 In case it happens again while doing some other experiments, can we call you or there is something we can try/save?

**Comment by davidetome** [comment]
Ciao @S-Dafarra , I simply cycled some position moving most of the joint of the `left-arm` using the sequence posted above in the `yarpmotorgui` to see if the board `EB24` disappear. 
Honestly, I didn't know that this happens on both the arms,  so this changes a little bit the focus.
 We could consider asking an opinion also to `firmware` guys (@marcoaccame @ale-git )...

cc @julijenv

**Comment by S-Dafarra** [comment]
> Honestly, I did not understand that this happens on both the arms, so this changes a little bit the focus.

I probably should have changed the title. I mentioned in a few comments above it was happening on both arms. Unfortunately, I was able to get a usable log only for the left arm. Most of the time it happens after an intensive session, where the log is already saturated since a while.

**Comment by S-Dafarra** [comment]
Interestingly, this happened once also on iCub3. So, I guess it is not a problem specific to ``iCubGenova04``.

**Comment by julijenv** [comment]
I dont have too much time to deal with this specific issue right now. But at the moment as you define it as non machine specific I would consider it a software issue. and I will let more competent people try to figure out, if it is software related or not before investigated too much time uselessly.
@pattacini

**Comment by S-Dafarra** [comment]
I don't know if it is related, but we noticed a similar behavior also for the neck joints. From time to time, the neck joints cannot be controlled anymore and we have to reboot the motors. I suspect that the issue is the same, and I think it has to do with the ``mc4plus`` boards that control all these joints. 

@julijenv @pattacini let me know if you prefer me to change the title of this issue or to open a new one.

cc @DanielePucci

**Comment by pattacini** [comment]
>let me know if you prefer me to change the title of this issue or to open a new one.

Title updated! Better off sticking to this issue to keep an eye on the long history.
In this sprint, @davidetome will get in touch w/ you to see how to proceed with this.

**Comment by S-Dafarra** [comment]
> > let me know if you prefer me to change the title of this issue or to open a new one.
> 
> Title updated! Better off sticking to this issue to keep an eye on the long history.
> In this sprint, @davidetome will get in touch w/ you to see how to proceed with this.

Ok! Notice that we don't have a lot of time. Starting from next week we will be involved in some shootings outside the lab. It would make our life much easier if we are able to face this during this week. 

At this point, I suspect that the problem is mainly related to firmware as it involves several boards on several robots. The nice thing is that we can use iCub3 for the tests even if we keep the greeny busy.

I don't know if it is related, but I remember a bug on the FT boards that caused them to get stuck because of communication issues: https://github.com/robotology/icub-firmware/issues/102

Is it possible that a similar bug is also present in the ``mc4plus`` firmware?

**Comment by pattacini** [comment]
We'll see how we can handle that, although it's difficult to promise to sort out a longstanding problem in a few days.

cc @davidetome @ale-git

**Comment by davidetome** [comment]
Ciao @S-Dafarra , if it is feasible me and @ale-git would be on this on Thursday morning(at least) to do some test.. is `iCubGenova09` available?

cc @pattacini @DanielePucci

**Comment by DanielePucci** [comment]
I further checked with @S-Dafarra and it is ok on Thursday! Please just book the robot

CC
@robotology/iit-dynamic-interaction-control

**Comment by DanielePucci** [comment]
CC @ale-git @davidetome @pattacini

**Comment by davidetome** [comment]
> Interestingly, this happened once also on iCub3. So, I guess it is not a problem specific to `iCubGenova04`.

@S-Dafarra since we're going to use iCub3 this morning. on which joint did it happened?

cc @ale-git

**Comment by pattacini** [comment]
From the OP:

>joints starting from the wrist prono-supination cannot be controlled and are kind of "compliant"

plus:

>We noticed a similar behavior also for the neck joints. From time to time, the neck joints cannot be controlled anymore and we have to reboot the motors. I suspect that the issue is the same, and I think it has to do with the mc4plus boards that control all these joints.

**Comment by davidetome** [comment]
@pattacini , I've read but I was asking if on iCub3 it happened on the neck, on the wrist or on both of them

**Comment by S-Dafarra** [comment]
> @pattacini , I've read but I was asking if on iCub3 it happened on the neck, on the wrist or on both of them

On iCub3 we noticed only the wrists for the time being.

**Comment by S-Dafarra** [comment]
In any case, let me stress again some of the things we noticed:
- we never managed to reproduce it properly, but lately is happening quite often. It seems to depend on the number of commands we give to the specific part. For example, it seems to happen more often when we use the hands, but it happens also when doing several yogas, when the wrists are not touched at all.
- the boards stop responding. The robot interface starts complaining that it did not have any reply from the board, and, clearly, the values in the motorgui appear to be fixed.
- On the iCubGenova04, it seemed to be happening on the neck when controlling the eyelids for example, so I suppose it is not matter of the power required to the boards.

**Comment by davidetome** [comment]
We moved the robot for a while trying (in particular the left forearm) to replicate the fault but it didn't happen.
Anyway looking again at the logs with @ale-git we agree that it should be due to an HW problem since the boards restart.
We agreed with @S-Dafarra to try changing the `mc4plus` controlling the forearms (EB24 and EB27) of the iCubGenova04 and in the meanwhile check wirings both EH and Power.

![output](https://user-images.githubusercontent.com/6638215/108359713-a2105080-71f0-11eb-8ceb-963ffcf1a2e1.gif)


cc @pattacini @DanielePucci @maggia80

**Comment by DanielePucci** [comment]
If the output of the analysis is to change the `mc4plus`, let us know when you would like to proceed with the substitution

**Comment by pattacini** [comment]
Hi @DanielePucci 

The availability of the boards and when to do this replacement needs to be discussed internally w/ @Fabrizio69 and @julijenv first.
Stay tuned.

**Comment by S-Dafarra** [comment]
The problem of planning this after the demo, is that the usage of the robot will drop. For iCubGenova04 in particular, the forearms might be removed for a while as well. I am just afraid that a late replacement may give us an answer on whether it worked or not with a lot of delay.

**Comment by maggia80** [comment]
I remember there was a timestamp in the eth packets sent by the boards. If the board resets we may see the timestamp restarting from 0.

**Comment by maggia80** [comment]
the other issue might be errors in the eth packets. I don't know if it's always on the diagnostic for such errors.

**Comment by S-Dafarra** [comment]
> I remember there was a timestamp in the eth packets sent by the boards. If the board resets we may see the timestamp restarting from 0.

But I think that we do not receive anything from those boads, so do we receive those packets?

I remember that we checked with FirmwareUpdater after that happened and the corresponding board was in the IDLE state.

**Comment by julijenv** [comment]
Hi everyone:
Just as a matter of sum up, The tests that produce yesterday @davidetome and @ale-git , was already done also on iCubGenova07 with the similar ending, no problem encountered.
I agree that the MC4+ has never been tested by the supplier but I personally doubt about an HW problem - this is why I ask you guys, .
Having talked this morning with @maggia80 , will hope to be able to have more insights from the board timestamp to understand if it resets or not - pehaps @marcoaccame can help us in this. But in the meanwhile we agreed that, asa we will be back from the next event we will try to change one board and keep under strict control to find out if changing the board prevent the forearm to IDLE.
@pattacini , @S-Dafarra , @kouroshD , @DanielePucci

**Comment by marcoaccame** [comment]
Hi all, I will have a look at the issue later today.

**Comment by marcoaccame** [comment]
here are some sparse preliminary thoughts. i will collect them later on.

In [here](https://github.com/robotology/icub-tech-support/issues/673#issuecomment-779739106), @S-Dafarra wrote:
>I don't know if it is related, but I remember a bug on the FT boards that caused them to get stuck because of communication issues: robotology/icub-firmware#102
Is it possible that a similar bug is also present in the mc4plus firmware?

The `mc4plus` does not use the above mentioned CAN driver. The motion control is done by the `mc4plus` w/out using CAN communication as the actuation terminates on its PWM ports. The communication which is involved is the UDP communication inside the ETH internal network. 

Having said that, thanks @S-Dafarra ... any hypothesis will be investigated.

**Comment by marcoaccame** [comment]
Analisis of log of july 2020 as in here.


1. crc error found
   
in the log of july 2020 i have spotted several  CRC errors from board 10.0.1.26 detected in its input port.  that causes the drop of the UDP frame in error w/  several consequences. The first of them is that several otehr boards complain because they detect some missing UDP frames. Further analysis later.  

```
490 1260.241039 <ERROR>  from BOARD 10.0.1.26 (left_arm-eb26-j12_15), src LOCAL, adr 0, time 1276s 569m 33u: 
(code 0x06000002, par16 0x0000 par64 0x3000000000000001) 
-> ETH monitor: detected RX CRC error in port specified by par16 
(0 = ETH input (P2/P13/J4) , 1 = ETH output (P3/P12/J5) , 2 = internal).  
Application state is in most significant nibble of par64: 0 -> N/A, 1 -> idle, 3 -> running. 
Number of errors in par64&0xffffffff + .

```

2. the board 10.0.1.24 disappears

the service `EthMonitorPresence` which runs on icub-head detects the absence of one board, the 10.0.1.24, which stays absent for more than 60 seconds.

```
5880 1360.385314 <DEBUG> eth::EthMonitorPresence: BOARD 10.0.1.24 (left_arm-eb24-j4_7) has been silent for 0.020087 sec (its timeout is 0.02 sec) 
```


This board stays silent for another 60 seconds.

```
6028 1420.389001 <DEBUG> eth::EthMonitorPresence: BOARD 10.0.1.24 (left_arm-eb24-j4_7) has been silent for another 60.0037 sec, for a total of 60.0238 sec 
```



3. all the ETH boards of the robot disappear

in the log of july 2020 i have spotted that the service `EthMonitorPresence` which runs on icub-head detects the absence of mostly every board inside the robot.

```
6084 1444.767627 <DEBUG> eth::EthMonitorPresence: BOARD 10.0.1.5 (torso-eb5-j0_2) has been silent for 0.0200102 sec (its timeout is 0.02 sec) 
 
6085 1444.767651 <DEBUG> eth::EthMonitorPresence: BOARD 10.0.1.6 (left_leg-eb6-j0_3) has been silent for 0.0200291 sec (its timeout is 0.02 sec) 
 
6086 1444.767655 <DEBUG> eth::EthMonitorPresence: BOARD 10.0.1.7 (left_leg-eb7-j4_5) has been silent for 0.0201232 sec (its timeout is 0.02 sec) 
 
6087 1444.767659 <DEBUG> eth::EthMonitorPresence: BOARD 10.0.1.27 (right_arm-eb27-j4_7) has been silent for 0.0200164 sec (its timeout is 0.02 sec) 
 
6088 1444.767663 <DEBUG> eth::EthMonitorPresence: BOARD 10.0.1.29 (right_arm-eb29-j12_15) has been silent for 0.0200527 sec (its timeout is 0.02 sec) 
 
6089 1444.777712 <DEBUG> eth::EthMonitorPresence: BOARD 10.0.1.1 (left_arm-eb1-j0_3) has been silent for 0.0200806 sec (its timeout is 0.02 sec) 
 
6090 1444.777730 <DEBUG> eth::EthMonitorPresence: BOARD 10.0.1.3 (right_arm-eb3-j0_3) has been silent for 0.0249074 sec (its timeout is 0.02 sec) 
 
6091 1444.777735 <DEBUG> eth::EthMonitorPresence: BOARD 10.0.1.8 (right_leg-eb8-j0_3) has been silent for 0.0249684 sec (its timeout is 0.02 sec) 
 
6092 1444.777740 <DEBUG> eth::EthMonitorPresence: BOARD 10.0.1.9 (right_leg-eb9-j4_5) has been silent for 0.0249648 sec (its timeout is 0.02 sec) 
 
6093 1444.777744 <DEBUG> eth::EthMonitorPresence: BOARD 10.0.1.11 (right_leg-eb11-skin) has been silent for 0.0249219 sec (its timeout is 0.02 sec) 
 
6094 1444.777748 <DEBUG> eth::EthMonitorPresence: BOARD 10.0.1.20 (head-eb20-j0_1) has been silent for 0.0248933 sec (its timeout is 0.02 sec) 
 
6095 1444.777751 <DEBUG> eth::EthMonitorPresence: BOARD 10.0.1.21 (head-eb21-j2_5) has been silent for 0.0202 sec (its timeout is 0.02 sec) 
 
6096 1444.777755 <DEBUG> eth::EthMonitorPresence: BOARD 10.0.1.22 (face-eb22-j0_1) has been silent for 0.0202057 sec (its timeout is 0.02 sec) 
 
6097 1444.777759 <DEBUG> eth::EthMonitorPresence: BOARD 10.0.1.25 (left_arm-eb25-j8_11) has been silent for 0.0251045 sec (its timeout is 0.02 sec) 
 
6098 1444.777763 <DEBUG> eth::EthMonitorPresence: BOARD 10.0.1.26 (left_arm-eb26-j12_15) has been silent for 0.0250769 sec (its timeout is 0.02 sec) 
 
6099 1444.777766 <DEBUG> eth::EthMonitorPresence: BOARD 10.0.1.28 (right_arm-eb28-j8_11) has been silent for 0.0251465 sec (its timeout is 0.02 sec)
```

board 10.0.1.24 was already lost some time earlier and board 10.0.1.10 is not activated by xml files.

Why all boards? It may be that the ETH goes down, maybe because the cable which goes from icub-head to board 10.0.1.20, the first of the chain.

After that, ... yarprobotineterfce cannot transmit anymore to ETH boards. See messages after.

```
6111 1445.000858 <ERROR> HostTransceiver::addSetROP__(): ERROR in eo_transceiver_OccasionalROP_Load() for BOARD w/ IP 0.0.1.1 after all attempts with id:  ID32 = 0x01000206 -> IND = 2, TAG = eoprot_tag_mc_joint_config_impedance 

etc.
```

**Comment by marcoaccame** [comment]
@julijenv, @S-Dafarra: do you have any recent log files?

**Comment by S-Dafarra** [comment]
> @julijenv, @S-Dafarra: do you have any recent log files?

Most of the times, the logger gets saturated before. There is a correlation between the logger being saturated and this error because it often happens after we use it for a long time and the total number of messages easily goes above 10000.

> all the ETH boards of the robot disappear

But then, why the rest of the robot is still functional?

**Comment by DanielePucci** [comment]
The issue occurred just few minutes before the demo in one of the very bad mikents. Forced to reboot.

![image](https://user-images.githubusercontent.com/5045846/108622285-fce7b900-7437-11eb-97ec-d47694247b32.jpeg)

@marcoaccame @maggia80

**Comment by DanielePucci** [comment]
At 6.00 p.m. and after 3 hour of continuous shooting only in the afternoon, the problem occurred again, this time on the head. The board was not pingable, and it seems that there was a traffic problem since the cameras were about to get stuck.

![image](https://user-images.githubusercontent.com/5045846/108632480-73ec7400-746f-11eb-983c-0c8a876f3ea6.jpeg)

**Comment by DanielePucci** [comment]
We tried to reboot the robot (only motors) four times, but whole-body dynamics device was not starting - we checked the logger but nothing particularly strange was happening, no error, although there were many due to other current hw problems so we are nut sure.

So we decided to reboot the head. After reboot though, wbd device kept not streaming. Suddenly, after additional time (~five minutes) the wbd device restarted streaming data.

@marcoaccame @maggia80

**Comment by DanielePucci** [comment]
Unfortunately, after we restarted, we were able to run the walking module only once. At the second time, the  walking module blocked when trying to connect to the torso, nor the yarpmotorgui was starting for the same reason. We are not sure if this latter problem (torso stuck) is actually related to this issue. I'm reporting here to detail the sequence of events after one instance of the issue occurrence.

**Comment by traversaro** [comment]
> 
> 
> We tried to reboot the robot (only motors) four times, but whole-body dynamics device was not starting - we checked the logger but nothing particularly strange was happening, no error (there were many other errors due to other current hw problems so we are nut sure).
> 
> So we decided to reboot the head. After reboot though, wbd device kept not streaming. Suddenly, after additional time (~five minutes) the wbd device restarted streaming data.
> 
> @marcoaccame @maggia80

FYI this seems an instance of the problem reported and tracked in https://github.com/robotology/icub-tech-support/issues/999 .

**Comment by DanielePucci** [comment]
Just for the record, the robot finally crashed without the crane in the very last sequence we were left to shoot before the occurrence of the issue.

We are now attempting fixes with @julijenv and @violadelbono  ![image](https://user-images.githubusercontent.com/5045846/108640789-e83d0c80-749b-11eb-9467-8a8b1964b951.jpeg)

We sincerely do not know what happened, probably the wbd being stuck sent wrong data to the walking controller and generated ill posed desired com velocity. We implemented a possible workaround here https://github.com/robotology/walking-controllers/commit/f43685800e48234e9eaf44f3d61c75f614f4d55e

**Comment by S-Dafarra** [comment]
Here a recent log. This is actually from yesterday morning, corresponding to the picture of https://github.com/robotology/icub-tech-support/issues/673#issuecomment-782834607, but I managed to upload it only now.

[log_icub-head_yarprobotinterface_6021_boardsStuck.txt](https://github.com/robotology/icub-tech-support/files/6025016/log_icub-head_yarprobotinterface_6021_boardsStuck.txt)

**Comment by marcoaccame** [comment]
> Here a recent log. This is actually from yesterday morning, corresponding to the picture of #673 (comment), but I managed to upload it only now.

Hi @S-Dafarra. Did you manage to check the status of board `10.0.1.24`? I mean:
- did it respond to `ping 10.0.1.24`?
- if so, what was its status in `FirmwareUpdater` during the execution of `yarprobotinterace`? `idle` or `running`?

I am asking because in this log file as in the previous one everything starts with board `10.0.1.24` (and only this board) that disappear to the monitoring service run by `yarprobotinterface`. See lines: `eth::EthMonitorPresence: BOARD 10.0.1.24 (left_arm-eb24-j4_7) has been silent for 0.0200384 sec (its timeout is 0.02 sec)` and it does not reappear anymore.
And if the board is not alive (or in `running` mode) than it cannot work properly.

To this board `10.0.1.24` could have happened one of two things:
1. a crash w/out a restart. In such a case it cannot be pinged anymore, nor be seen by `FirmwareUpdater`.
2. a reset of the board, probably due to electrical reasons, w/ a restart. In such a case, the ping will succeed and `FirmwareUpdater` will show the board in `idle` state.

**Comment by S-Dafarra** [comment]
> > Here a recent log. This is actually from yesterday morning, corresponding to the picture of #673 (comment), but I managed to upload it only now.
> 
> Hi @S-Dafarra. Did you manage to check the status of board `10.0.1.24`? I mean:
> 
>     * did it respond to `ping 10.0.1.24`?
> 
>     * if so, what was its status in `FirmwareUpdater` during the execution of `yarprobotinterace`? `idle` or `running`?
> 
> 
> I am asking because in this log file as in the previous one everything starts with board `10.0.1.24` (and only this board) that disappears to the monitoring service run by `yarprobotinterface`. See lines: `eth::EthMonitorPresence: BOARD 10.0.1.24 (left_arm-eb24-j4_7) has been silent for 0.0200384 sec (its timeout is 0.02 sec)` and it does not reappear anymore.
> And if the board is not alive (or in `running` mode) then it cannot work properly.
> 
> To this board `10.0.1.24` could have happened one of two things:
> 
>     1. a crash w/out a restart. In such a case it cannot be pinged anymore, nor be seen by `FirmwareUpdater`.
> 
>     2. a reset of the board, probably due to electrical reasons, w/ a restart. In such a case, the ping will succeed and `FirmwareUpdater` will show the board in `idle` state.

Contrarily to what I remembered, immediately after the board was not pingable neither visible from the ``FirmwareUpdater``, so I think we are in case 1

**Comment by marcoaccame** [comment]
> Contrarily to what I remembered, immediately after the event the board was not pingable neither visible from the FirmwareUpdater, so I think we are in case 1

OK, that is also confirmed by the failure of asking the `controlmodestatus` of joints on that board. If the board is alive the board must reply.

`theNVmanager::Impl::ask() had a timeout for BOARD left_arm-eb24-j4_7 IP 10.0.1.24 and nv ID32 = 0x0100000c -> IND = 0, TAG = eoprot_tag_mc_joint_status_core_modes_controlmodestatus`

**Comment by marcoaccame** [comment]
@S-Dafarra, I can see that the EXT fault was pressed. Do you confirm that someone pressed it?

**Comment by pattacini** [comment]
>1. a crash w/out a restart. In such a case it cannot be pinged anymore, nor be seen by FirmwareUpdater.

- Is the board `10.0.1.24` of MC4Plus type?
- What could cause such a permanent crash?

**Comment by marcoaccame** [comment]
> - Is the board `10.0.1.24` of MC4Plus type? 

**YES**.

> - What could cause such a permanent crash?

don't know yet.

I am adding an extra question:
- the dual of `10.0.1.24` is the `10.0.1.27`. it should do similar activity but on the right arm. Why in two logs we saw problems only on the `10.0.1.24`?

I will look at possible differences between the two.

**Comment by S-Dafarra** [comment]
> @S-Dafarra, I can see that the EXT fault was pressed. Do you confirm that someone pressed it?

Not in that moment for sure. 

In the previous run, we forgot to plug the fault. We plugged it, restarted the robot interface and we still received some of those messages, but the robot started cleanly. At least this is what @isorrentino and I remember. In any case, the robot was not faulted since a long time.

**Comment by S-Dafarra** [comment]
> > - Is the board `10.0.1.24` of MC4Plus type? 
> 
> **YES**.
> 
> > - What could cause such a permanent crash?
> 
> don't know yet.
> 
> I am adding an extra question:
> - the dual of `10.0.1.24` is the `10.0.1.27`. it should do similar activity but on the right arm. Why in two logs we saw problems only on the `10.0.1.24`?
> 
> I will look at possible differences between the two.

It did not happen in this case, but it often happens also on the right arm. Very rarely on both arms at the same time.

**Comment by marcoaccame** [comment]
The boards `10.0.1.24` and `10.0.1.27` are dual board of type `mc4plus`. 

They are similar and:

- run the same FW version (latest master: ver 3.26), 
- have the same settings for control loop and TX rate of regulars (every 3 ms).
- they offer the same services configured w/ the same parameters.
  - motion control for joints 4 to 7 of the arms (left -> .24, right -> .27)
  - skin streaming.

I cannot see any reason why the `10.0.1.24` stops responding and the `10.0.1.27` does not, but so far I have seen in the logs only the disappearance of the `10.0.1.24`.  

So far, I have analyzed four logs for iCubGenova04. All of them inside this issue and dated: july 2018, october 2019, july 2020, and 23 feb 2021.

In all of them, the `EthMonitorPresence` module always detects the disappearance of the board `10.0.1.24`.  Only this board and never other boards. And surely it is this disappearance which causes the unresponsiveness of the left arm.

On the robot I have also spotted some errors typically associated with electrical noise. They are not responsible of disappearance of the 10.0.1.24 but they surely tell about electrical noise in the minutes around this disappearance.  


On the left arm, mc4plus boards 10.0.1.24 and 10.0.1.26:

- spike errors of the AEA encoder on the board 10.0.1.24  (log of 23 feb 2021). Probably not related with its disappearance of 10.0.1.24 because spikes happen 5 minutes earlier.
- CRC errors on the board 10.0.1.26 (log of july 2020) @ second 1260-62. Probably not related with its disappearance of 10.0.1.24 because CRC errors stop 98 seconds earlier.

On the right arm, mc4plus board 10.0.1.27:

- spike errors of the AEA encoder on the board 10.0.1.27  (log of july 2020). One spike just 17 seconds after the disappearance of 10.0.1.24.

On the head, mc4plus board 10.0.1.21 (moves the neck-yaw + eyes):

- spike errors of the AEA encoder on the board 10.0.1.21  (log of july 2020). One spike just 17 seconds after the disappearance of 10.0.1.24.


**Temporary conclusion (until new evidence arises )**

Until new evidence in form of logs may suggest something else, my idea is that it can be either:

- a faulty board as already suggested by @davidetome in [here](https://github.com/robotology/icub-tech-support/issues/673#issuecomment-781322843), or
- or a very nasty problem due to electrical noises which for that particular board, maybe also partly faulty, cause the freeze of the MCU and not of the ETH switch chip. 

I would surely substitute the board and observe what happens.

**Comment by S-Dafarra** [comment]
We experience the same problem also with iCub3. I think that is not considered in the temporary conclusion.

The fact that we managed to retrieve logs only when it happened on the left is, in my opinion, just a coincidence.

Finally, notice that we had a similar behavior few hours later on the board that controls the neck pitch and roll (see https://github.com/robotology/icub-tech-support/issues/673#issuecomment-782889967). That ultimately cause the robot to fall.

Please, please, please, please, can we rule out the electrical failure for a second, and assume for once that there is something that has to do with how these boards run? Then, i  the end we may conclude that there is indeed a cable broken, but until now I have the impression we are affected by the ["Street-light effect"](https://en.wikipedia.org/wiki/Streetlight_effect)

**Comment by julijenv** [comment]
> The boards `10.0.1.24` and `10.0.1.27` are dual board of type `mc4plus`.
> 
> They are similar and:
> 
> * run the same FW version (latest master: ver 3.26),
> * have the same settings for control loop and TX rate of regulars (every 3 ms).
> * they offer the same services configured w/ the same parameters.
>   
>   * motion control for joints 4 to 7 of the arms (left -> .24, right -> .27)
>   * skin streaming.
> 
> I cannot see any reason why the `10.0.1.24` stops responding and the `10.0.1.27` does not, but so far I have seen in the logs only the disappearance of the `10.0.1.24`.
> 
> So far, I have analyzed four logs for iCubGenova04. All of them inside this issue and dated: july 2018, october 2019, july 2020, and 23 feb 2021.
> 
> In all of them, the `EthMonitorPresence` module always detects the disappearance of the board `10.0.1.24`. Only this board and never other boards. And surely it is this disappearance which causes the unresponsiveness of the left arm.
> 
> On the robot I have also spotted some errors typically associated with electrical noise. They are not responsible of disappearance of the 10.0.1.24 but they surely tell about electrical noise in the minutes around this disappearance.
> 
> On the left arm, mc4plus boards 10.0.1.24 and 10.0.1.26:
> 
> * spike errors of the AEA encoder on the board 10.0.1.24  (log of 23 feb 2021). Probably not related with its disappearance of 10.0.1.24 because spikes happen 5 minutes earlier.
> * CRC errors on the board 10.0.1.26 (log of july 2020) @ second 1260-62. Probably not related with its disappearance of 10.0.1.24 because CRC errors stop 98 seconds earlier.
> 
> On the right arm, mc4plus board 10.0.1.27:
> 
> * spike errors of the AEA encoder on the board 10.0.1.27  (log of july 2020). One spike just 17 seconds after the disappearance of 10.0.1.24.
> 
> On the head, mc4plus board 10.0.1.21 (moves the neck-yaw + eyes):
> 
> * spike errors of the AEA encoder on the board 10.0.1.21  (log of july 2020). One spike just 17 seconds after the disappearance of 10.0.1.24.
> 
> **Temporary conclusion (until new evidence arises )**
> 
> Until new evidence in form of logs may suggest something else, my idea is that it can be either:
> 
> * a faulty board as already suggested by @davidetome in [here](https://github.com/robotology/icub-tech-support/issues/673#issuecomment-781322843), or
> * or a very nasty problem due to electrical noises which for that particular board, maybe also partly faulty, cause the freeze of the MCU and not of the ETH switch chip.
> 
> I would surely substitute the board and observe what happens.

WHen we will be back in Genvoa we will change the board IP 24 and see if the errors stop arising for this specific board we will later change also 27 and 21 if just both of them keep messing around.
My best suggestion is doing one at a time just to remove every once of doubt.
thx @marcoaccame

**Comment by S-Dafarra** [comment]
This morning it happened again, this time on the head. We also recently started the yarprobotinterface, so we finally managed to have a usable log. Note that this happened just few minutes (2~3) before some shootings and I managed to notice it just for luck.

Also note that the robot was completely still, hanging. As it happens for the arm, the board 10.0.1.20 is no more pingable and not visible from the Firmware updater. The neck pitch and neck roll joints are not as "soft" as in the idle case, but not rigid at all. We can still move them by hand, but then they remain kind of still, like if a constant PWM is applied to them.

Here the log 
[log_icub-head_yarprobotinterface_1926_headStuck.txt](https://github.com/robotology/icub-tech-support/files/6037735/log_icub-head_yarprobotinterface_1926_headStuck.txt)

**Comment by marcoaccame** [comment]
Thanks @S-Dafarra for the new log. 

I can see that the board .20 disappears. This phenomenon is what triggers the problems and MUST be understood.
The fact that the motors are stuck is also useful information. I also have  a photo of `FirmwareUpdater` that I attach (courtesy of @maggia80).

![firmware-updater-eb20](https://user-images.githubusercontent.com/7148284/109125546-793a0f00-774c-11eb-8a8d-40281080e852.png)

**Comment by marcoaccame** [comment]
So, what happens? So far, **sadly**, I haven't seen in the log file any resolutive information.

**To find the cause of the problem I believe** it is necessary to add some focused diagnostic messages which tell us the status of the board when the problem comes. I have some ideas which we can discuss in more details in the next days.  One idea is that when the board freezes we can use the internal watchdog timer (WDT) to restart it in such a way that it transmits the cause of the problem.

**Comment by S-Dafarra** [comment]
> So, what happens? So far, **sadly**, I haven't seen in the log file any resolutive information.
> 
> **To find the cause of the problem I believe** it is necessary to add some focused diagnostic messages which tell us the status of the board when the problem comes. I have some ideas which we can discuss in more details in the next days.  One idea is that when the board freezes we can use the internal watchdog timer (WDT) to restart it in such a way that it transmits the cause of the problem.

That would be great!

**Comment by S-Dafarra** [comment]
When https://github.com/robotology/icub-tech-support/issues/673#issuecomment-782834607 happened, I did this video. The wrist was supposed to be straight


https://user-images.githubusercontent.com/18591940/109349548-6d4a6c00-7876-11eb-9c82-f5749be46a28.mp4

**Comment by marcoaccame** [comment]
Hi all, late Friday 19 March I tested on the robot `iCubGenova04` a version of binaries for the `ems` and `mc4plus` which are immune to multiple start / stop of `yarprobotinterface`. 

The tests were **successful**: I run cycles of > 20 times of start / stop and no disappearance of boards was found. I also left `yarprobotinterface` running for about 30 minutes. 

On the other hand, the released binary versions now in master / devel caused disappearance after just 2/3 cycles.

Since then, I have produced two PRs: one for [sources](https://github.com/robotology/icub-firmware/pull/176) and one for [binaries](https://github.com/robotology/icub-firmware-build/pull/25) which I will soon merge into `robotology/devel`. 

I would like to test the binaries once more time since I had to rebase my branch vs latest changes in devel.

**Comment by pattacini** [comment]
As reported in https://github.com/robotology/icub-firmware/issues/174#issuecomment-806590647, the problem should be solved 🎉

Closing, feel free to open it up again if required.

**Comment by pattacini** [comment]
As reported in https://github.com/robotology/icub-firmware/issues/174#issuecomment-806590647, the problem should be solved 🎉

Closing, feel free to open it up again if required.

**Comment by S-Dafarra** [comment]
Unfortunately, this happened again. It was probably the second or the third time that I launched the ``yarprobotinterface`` after rebooting the motors. The robot was on since a couple of hours (so no usable log). When I was about to switch it off, I noticed that left arm did not close correctly, and the wrist is in the usual "half-controlled" state. I run ``FirmwareUpdater`` and the ``10.0.1.24`` board is not there
![image](https://user-images.githubusercontent.com/18591940/113281392-c5491800-92e5-11eb-8920-1a8484168887.png)

**Comment by pattacini** [comment]
cc @marcoaccame

**Comment by marcoaccame** [comment]
ok, I have seen it. I will prepare special log messages which will be logged by the mc4plus just before the board disappears and sent after a reboot so that we can exactly see which is the critical point.

**Comment by marcoaccame** [comment]
I will reopen https://github.com/robotology/icub-firmware/issues/174

**Comment by Nicogene** [comment]
We are experiencing the very same issue on `iCubErzelli02`

cc @vvasco @vtikha

**Comment by DanielePucci** [comment]
@Nicogene it would be nice to link the issue in iCub support where you reported the problem of `iCubErzelli02`

**Comment by Nicogene** [comment]
We didn't open a dedicated issue, but we can open it if necessary

**Comment by traversaro** [comment]
I added the `iCubErzelli02` label to this issue.

**Comment by S-Dafarra** [comment]
This is still happening from time to time. It happened on ``iCubGenova04`` during the preparation of a demo, affecting the eyes vergence.

It also happened a couple of times recently on iCubGenova09 affecting mainly the neck and the wrist pronosupination joints.

**Comment by DanielePucci** [comment]
CC @marcoaccame @pattacini

**Comment by pattacini** [comment]
Yes, we're aware of this, but we gave priority to the other points of the list.
In this respect, @marcoaccame did work on the new level of diagnostics for the problem with the AMO readouts (see https://github.com/robotology/icub-firmware/pull/188).

Starting next week, we will be able to get back to this.

**Comment by S-Dafarra** [comment]
> Yes, we're aware of this, but we gave priority to the other points of the list.
> In this respect, @marcoaccame did work on the new level of diagnostics for the problem with the AMO readouts (see [robotology/icub-firmware#188](https://github.com/robotology/icub-firmware/pull/188)).
> 
> Starting next week, we will be able to get back to this.

Yes, the problem with the AMO has the priority. I commented mainly for historical purposes

**Comment by DanielePucci** [comment]
Starting from September 24th, we will be having three important demos outdoor. One of these will be for an institutional event in Rome where the Prime Minister will be present - issues in https://github.com/dic-iit/lab-events-demos will follow soon.

It would be important then to attempt a definite solution to this issue by September 24th

CC
@marcoaccame @maggia80

**Comment by pattacini** [comment]
Following up on @marcoaccame suggestion, we decided to do the following:
- make sure that iCub 3 will have the latest Distro 2021.08 FW for the MC4Plus.
- make the current official FW even more verbose in terms of diagnostic as we have now more RAM to dedicate to the log.
- while awaiting feedback and logs to analyze (there's a clear reproducibility problem), we may think to apply a recovery strategy to let the board restart in a bunch of ms with all the motion control layers already configured.

**Comment by DanielePucci** [comment]
This is a gentle reminder for this issue in light of the looming important demos we will have soon (e.g. https://github.com/ami-iit/lab-events-demos/issues/186, where international guests are expected)

CC
@pattacini @marcoaccame @maggia80

**Comment by DanielePucci** [comment]
I did not notice the yesterday's @pattacini comment due to browser window dimension settings, sorry for that. Let me just add that the demos will be done with iCubGenova04. Do not hesitate to update us in the case of additional news

**Comment by pattacini** [comment]
pushed code in https://github.com/marcoaccame/icub-firmware/tree/feat/mpucontext-on-fatalerror
now the ems, mc4plus, mc2plus after a fatal error send diagnostic messages w/ content of the main [system control registers](https://developer.arm.com/documentation/ddi0439/b/System-Control/Register-summary) and if the error is caught by the `HardFault_Handler()` they also send info about the last executed instructions (in the form of the r1, r2, r3, etc registers)
 
the code is ready and the ems, mc4plus, mc2plus also have a new version number.

however, i have to perform tests yet:
- [ ] normal running on the board
- [ ] use of yarprobotinterface
- [ ] simulation of a fatal error and observation of the values of the registers

**Comment by DanielePucci** [comment]
For the record, the comment was written by @marcoaccame somewhere else (@S-Dafarra informed me about this)- it would be nice to add the pointer to the issue if accessible. 

@marcoaccame It seems to me that we are still in the debugging stage, so let's hope that the new diagnostics will help spot the problem soon since we are getting closer to the events outdoor

**Comment by pattacini** [comment]
Hi @DanielePucci 

@marcoaccame wrote that specific post into our internal workflow 10 minutes before our meetup yesterday. I copied out it soon afterward in this place as it's my responsibility to keep the interface up-to-date.

The pressure of the event is fairly clear to us, no doubt, and we're working hard to narrow down this critical problem but I don't think that continuous alerts may improve the situation.

**Comment by DanielePucci** [comment]
@pattacini 
> The pressure of the event is fairly clear to us, no doubt,  and we're working hard to narrow down this critical problem but I don't think that continuous alerts may improve the situation.

Ack, probably we can discuss offline about a roadmap so us alerts can reduce since we agree on a plan. Since the report in the internal flow was not pointed, then the action plan and associated priority was not clear. Probably we may also discuss on how to handle this information exchange in future similar situations .

**Comment by pattacini** [comment]
Ok @DanielePucci 👍🏻 

To clear things up, @marcoaccame is working 100% on this and the experimental FW is almost ready. We will provide info on this by tomorrow.

Also, as mentioned during the alignment meetup, we are going to use our setup (already available) in parallel with the robot, right with the aim to mitigate the critical aspects of timing since the problem occurs quite rarely.

cc @sgiraz

**Comment by DanielePucci** [comment]
Thanks a lot @pattacini for clarifying the timeline, this was the information a missed. Sorry for the annoying alerts

**Comment by marcoaccame** [comment]
> However, i have to perform tests yet:
> 
> * [x]  normal running on the board
> * [ ]  use of yarprobotinterface
> * [x]  simulation of a fatal error and observation of the values of the registers

the board is able to detect a hard fault induced by accessing non-existing RAM and to deliver the information over diagnostics messages. the information has proven satisfactory to detect the type of error and where in the code it happened.

we are close to a new release.

**Comment by pattacini** [comment]
iCub 3 is now running the latest MC4Plus FW sporting the augmented diagnostics: @davidetome tested it today.

The PR that will be merged (the robot is already up-to-date):
- https://github.com/robotology/icub-firmware-build/pull/35
- https://github.com/robotology/icub-firmware/pull/201

As anticipated, we are also preparing a setup where to put to test one MC4Plus in parallel.

**Comment by marcoaccame** [comment]
> The PR that will be merged (the robot is already up-to-date):

The two PRs (one binaries and one sources) have just been merged.

**Comment by pattacini** [comment]
With regards to our internal setup for the tests on the MC4Plus, find here below a brief update.

@marcoaccame accurately summarized the information we have on the problem.

<details>
<summary>🔘 Analysis of where the fatal error has happened on iCubGenova09</summary>

see https://github.com/robotology/icub-firmware/issues/174#issuecomment-896159522 for the only logs (so far) of two mc4plus boards which triggered the fatal error handler.

some more info on them is here:

**mc4plus @ 10.0.1.20, head-eb20-j0_1:** 

- see https://github.com/icub-tech-iit/fix/issues/951#issuecomment-913618978. it triggers the HardFault handler after 18613 seconds (about 5 hours and 6 minutes). 
- the board on the iCubGenova09 robot has the following general parameters:
  - https://github.com/robotology/robots-configuration/blob/master/iCubGenova09/hardware/electronics/head-eb20-j0_1-eln.xml. it sends up regular udp packets every 5 ms.
- it executes services of:
  - **motion control** as described by https://github.com/robotology/robots-configuration/blob/master/iCubGenova09/hardware/motorControl/head-eb20-j0_1-mc_service.xml. it moves two dc motors, reads two AEA encoders and two qenc motor encoders. The two joints are the pitch and the roll of the neck as described by https://github.com/robotology/robots-configuration/blob/master/iCubGenova09/hardware/mechanicals/head-eb20-j0_1-mec.xml. they are **coupled**. Stefano Dafarra refers that: `Ciao Marco! Si guarda a volte capita che succeda senza che il robot faccia alcun movimento mentre è in position control`.
  -  **nothing else**.
- **comment**: `very basic activity, 2 joints only, coupled, no CAN boards`.

mc4plus @ 10.0.1.24, left_arm-eb24-j4_7:

- the board triggers the hardFault handler after 25630045 ms = 25630 sec = 7.1 hours
- the board on the iCubGenova09 robot has the following general parameters: https://github.com/robotology/robots-configuration/blob/master/iCubGenova09/hardware/electronics/left_arm-eb24-j4_7-eln.xml. it sends up regulars every 3 ms.
- it executes services of:
  - **motion control** as described in https://github.com/robotology/robots-configuration/blob/master/iCubGenova09/hardware/motorControl/left_arm-eb24-j4_7-mc_service.xml. it moves four dc motors and reads mixed encoders types: aea, qenc, absanalog. the four joints are left wrist and one finger as described in https://github.com/robotology/robots-configuration/blob/master/iCubGenova09/hardware/mechanicals/left_arm-eb24-j4_7-mec.xml. **two joints are coupled**.
  - **skin** as described in https://github.com/robotology/robots-configuration/blob/master/iCubGenova09/hardware/skin/left_arm-eb24-j4_7-skin.xml and https://github.com/robotology/robots-configuration/blob/master/iCubGenova09/hardware/skin/left_arm-eb24-j4_7-skinSpec.xml. It uses 7 mtb4 boards on CAN bus
  - **inertials** as described by https://github.com/robotology/robots-configuration/blob/master/iCubGenova09/hardware/inertials/left_arm-eb24-j4_7-inertials.xml. It uses 7 mtb4 boards on CAN bus.
- **comment**: `complex activity, 4 joints, high traffic on CAN bus`.





**Conclusion**

On board 10.0.1.20 of the robot where we see the error handler being triggered the activity of the mc4plus is moderate. It runs two DC motors with the AEAs. The motors are coupled. No CAN activity. The problem happened when the joints are in control mode and not moved at all.

On board 10.0.1.24 we have more joints and CAN bus activity. We also have two coupled motors.

So far our setup is only one DC motor with one AEA. It is very similar to the one of board 10.0.1.20. In it we only miss the second motor and the coupling.
</details>

<details>
<summary>🔘 Some further useful information</summary>

I have prepared some analysis of the existing log files w/ the fatal errors and I have talked w/ @S-Dafarra to understand what sort of actuation they were doing on the boards. 

main outcome from him was: 
`Ciao Marco! Si guarda a volte capita che succeda senza che il robot faccia alcun movimento mentre è in position control`.

in any case, let me post separately the analysis and later some conclusions of mine.
</details>

## Our plan with the setup

>- **stage 1**: Do tests on our existing setup w/ one motor only + AEA, w/ control mode and no movement. No debugger attached. We look at the diagnostics messages. If fatal error happens, attempt again w/ the debugger attached and breakpoint positioned in the fatal error handler. In this way we can look at the full stack history and debug will be much easier..
>- **stage 2**: If the above test does not produce errors, attempt control mode w/ more and more elaborated movements. If teh fatal error happens, attach the debugger.
>- **stage 3**: If still no fatal error, ... improve teh setup so that it is equal to board 10.0.1.20: attach to teh mc4plus  a second motor + AEA and use the very same xml files of the board 10.0.1.20 on the robot w/ coupled motors.

## Preliminary results obtained by @sgiraz 

>Today the `Test 1` has been performed for 5 hours continuously.
Nothing seems to have happened.

[yarprobotinterface_5h_log.txt](https://github.com/robotology/icub-tech-support/files/7212405/yarprobotinterface_5h_log.txt)

**Comment by S-Dafarra** [comment]
While testing on the stage in Turin, this happened once. Here the log: 
[log_icub-head_yarprobotinterface_1630_lostYaw.txt](https://github.com/robotology/icub-tech-support/files/7235129/log_icub-head_yarprobotinterface_1630_lostYaw.txt)

Unfortunately, the firmware is not the latest, but the one from the latest release. We did not manage to update it before leaving.

It happened also another couple of times later, but the log was saturated.

**Comment by pattacini** [comment]
Thanks @S-Dafarra for reporting 👍🏻 
The Murphy law strikes back, always in the most unfortunate conditions.

While @marcoaccame can be having a look at the report anyway, you could flash the exp. FW with the latest diags.

On our side, we have started out w/ `test phase 2`:

>- The tests developed with the collaboration of @triccyx are running on mc4Plus. In summary, the test performs the following steps:
    - Start servers and interfaces
    - Move the joint with a degree pattern {-30,30,-60,60, ... } of 9 different positions
    - Close resources and stop server and interfaces
>- After 100 iterations of `Test 2` no fatal errors have occurred so far.

Thus, we are going to make more substantial experiments for `Test 2` and let it run for hours.

**Comment by marcoaccame** [comment]
> While testing on the stage in Turin, this happened once. Here the log:
> [log_icub-head_yarprobotinterface_1630_lostYaw.txt](https://github.com/robotology/icub-tech-support/files/7235129/log_icub-head_yarprobotinterface_1630_lostYaw.txt)

Hi @S-Dafarra, I am already looking at the log. thanks.

**Comment by marcoaccame** [comment]
According to this log, the board does not restart and stays unconnected from yarprobotinterface for at least 909 seconds until its shutdown.

It is the mc4plus at address 10.0.1.21, it has fw version v3.36 and even if it is not the latest one, with this version the fatal error handler forces a restart. 

Strange that it did not restart. On the two logs we have seen (on https://github.com/robotology/icub-firmware/issues/174#issuecomment-896159522) the two mc4plus boards have a similar version (v103.36 built on 2021 Jun 21 15:45 which is the testing version of the release 3.36) and they restart after having caught the fatal error.

I would recommend to update to the latest release and wait for new logs.
In the meantime I will do a deeper comparison to the three logs we have.

**Comment by marcoaccame** [comment]
> In any case I am now working to understand more and more. In particular, I am investigating:
> 
> * how to transform the error to have the flag PRECISERR set instead, so that we can read the faulting address in register BFAR with a successive happening of the bug.

I have just produced an experimental binary for the mc4plus which, according to ARM (see here), will produce a PRECISERR flag in the system register. 

We are going to validate it first on our setup and then we can move it for use in iCub.

**Comment by marcoaccame** [comment]
The latest log collected during the [Maker Faire](https://github.com/robotology/icub-tech-support/issues/673#issuecomment-939441529) had shown that the fatal error due to a bus fault happened inside the [RTOS](https://github.com/robotology/icub-tech-support/issues/673#issuecomment-940067128) and that the last scheduled thread was the Timer Manager. 

Also, there was the need to produce a fault w/ the PRECISERR bit set which allows us to collect more precise information about the memory location which produced the bus fault.

As a result of that we have produced a new binary version for the mc4plus which improves the log information with:
- PRECISERR hopefully happening
- additional trace information inside the thread Timer Manager and around the guilty RTOS function

More information about how to read the improved log information can be found in the PR of [icub-firmware](https://github.com/robotology/icub-firmware/pull/209).

The new binary is now in robotology's `icub-firmware-build` branch `devel `under the name of [mc4plus-trace-rtos.hex](https://github.com/robotology/icub-firmware-build/blob/devel/ETH/MC4PLUS/bin/application/mc4plus-trace-rtos.hex).

It was successfully tested on a dedicated setup which runs a motor via `yarprobotinterface` and I recommend to run it on `iCubGenova08` and `iCubGenova04`.

**Comment by marcoaccame** [comment]
I have produced a detailed report of the log which I enclose in a dedicated post.
In short, what come out is:
- we know what memory violation causes the fatal error: a dirty pointer;
- we need a further FW release to understand why the pointer gets dirty.

**Comment by marcoaccame** [comment]
## Report of the log

The log is the one reported in [here](https://github.com/robotology/icub-tech-support/issues/673#issuecomment-954624585).

The fatal error type is  a `Bus Fault` which happens when the scheduled thread is  the `Timer Manager`. It happens in the same place as reported by the previous log obtained at the Maker Faire: inside the RTOS function `rt_rmv_dly (P_TCB p_task)`.  So far, we have the same information as before. Even better: we have a confirmation.



> _Explanation_:
>
> By looking at the pc value (0x08057b80) and searching inside the .map file, I can say where in the codespace the fault handler was called: inside function *rt_rmv_dly()*. The type of fault and the last scheduled thread are just printed in the log: *IRQHan BusFault Thread tmrma*.



The additional information we have with this new log is that the error happens just after the thread Timer Manager is woken up by the scheduler of  the RTOS and attempts to retrieve the timer which has just expired.  The Timer Manager retrieves the timer with the function `osal_messagequeue_getquick()` which in turns calls other RTOS functions until one finally calls `rt_rmv_dly (P_TCB p_task)` with an invalid value of `p_task`. Inside this function an attempt to write at an invalid memory location triggers the fatal error handler.


> _Explanation_:
>
>  From the tracing of the Timer Manager we know that the Sys Tick handler has completed the execution of function `s_eom_timerman_OnExpiry()` which in line 229 alerts the Timer Manager thread to wake up and retrieve the expired timer. The value of the timer is OK at this point as it points to the valid RAM address `TM3 = 0x20007020`. The Timer Manager thread at this point should wake up and retrieve the value of the timer with the call `osal_messagequeue_getquick() `which goes inside the RTOS.  Its execution is confirmed by the analysis of the  RTOS tracing.
>
>  | ![wake-up-tmr-man](https://user-images.githubusercontent.com/7148284/140047060-c6260147-bce2-474e-a1e5-de1822fd713e.jpg) | ![tmr-man-wakes-up](https://user-images.githubusercontent.com/7148284/140047071-68d42be4-efd1-498d-89e6-520c22f09bdf.jpg) |
>  | ---- | ---- |
>
>  **Figure**. This is the code which anticipates the fault inside the RTOS. Line 229 on the left wakes up the Timer Manager thread. Line 1203 on the right is the function which is executed. We know that line 1204 is not executed because it points to function `s_eom_timerman_tskproc_forever()` which does not prints its log information. So, the fatal error must happen inside `osal_messagequeue_getquick()`.
>
>
>  The tracing of the RTOS tells us that `rt_rmv_dly()` assumes a value of `p_task` equal to `0x080002e1` which is very wrong. When this pointer is derefenced it produces the fatal error with an attempt to write to memory location `0x04e7fef3`. This address is seen inside the register `BFAR` and also verified with the arithmetic of pointers from the trace information we have. With reference to the picture and the log: `p_b` has value `RT3 = 0x04e7fee7` and `p_b->p_dlnk` has an offset of 12 bytes to produce `0x04e7fef3`. The error happens in line 290.
>
>  | ![fatal-error-line-290](https://user-images.githubusercontent.com/7148284/140039294-7d0ab46d-d30a-4f3f-b7b2-93ac36e14f97.jpg) | ![fatal-error-caller-line-1269](https://user-images.githubusercontent.com/7148284/140039322-78e6b513-79ee-418a-b561-d39060e9f354.jpg) |
>  | ---- | ---- |
>
>  **Figure**. The code which triggers the fatal error is at line 290 in the left picture, where there is an attempt to write an invalid memory location after having deferenced the pointer `p_task` passed as argument of function `rt_rmv_dly()`. We have also traced back the origin of the invalid pointer up to line 1256 of the picture in the right. Unfortunately our trace instructions don't go further backwards.
>
>  If we trace back using the log information we find out that `rt_rmv_dly()` was called inside function `iitchanged_rt_mbx_wait()` at line 1269 of the picture with the wrong argument obtained at line 1256 by function `p_TCB = rt_get_first ((P_XCB)p_MCB);`. We don't have trace information in here to understand why `p_TCB` is wrong. What we know from code inspection however is that the high level function inside the Timer Manager which triggers this code is `osal_messagequeue_getquick()`. 



Now that we have spotted the precise execution path which causes the fatal error we **have to add further trace information inside the RTOS** from the call of `osal_messagequeue_getquick()` down to the internals where the fault happens.

**Comment by marcoaccame** [comment]
> - [ ] produce a new FW which is able to log about the execution path and arguments and variables of a wider section of code

A new FW release able to trace the execution of `osal_messagequeue_getquick()` is under test and it will soon be released.

**Comment by S-Dafarra** [comment]
Great @marcoaccame! If it can help, we had another log: 
[log_icub-head_yarprobotinterface_2863_another_wristFailure.txt](https://github.com/robotology/icub-tech-support/files/7467754/log_icub-head_yarprobotinterface_2863_another_wristFailure.txt)

If I checked it correctly, with respect to the previous log, only the RT5 value is different.

**Comment by marcoaccame** [comment]
> If I checked it correctly, with respect to the previous log, only the RT5 value is different.

  It can be. The value in RT5 is obtained by de-referencing the address 
  contained in RT3 which has value 0x04e7fee7. Now, this address space
  is reserved or not used (see the picture w/ memory map of our micro) 
  and hence it can change.

![image](https://user-images.githubusercontent.com/7148284/140079131-00a1e9b1-ac06-4e74-b439-56c17bbde169.png)

**Comment by marcoaccame** [comment]
> - [ ] produce a new FW which is able to log about the execution path and arguments and variables of a wider section of code

The PR is ready. We need only to test the binary on our setup before we can release it for the robot.

**Comment by marcoaccame** [comment]
> - [ ] use the new FW
> - [ ] wait for another log

Hi @S-Dafarra, a new version (v 3.40) for the binary of the `mc4plus` board is available for use on the robot. It is named  `mc4plus-trace-rtos.hex` and is here: https://github.com/robotology/icub-firmware-build/tree/devel/ETH/MC4PLUS/bin/application. 

Pls use it and collect logs.

Thanks,

**Comment by isorrentino** [comment]
Hi @marcoaccame, yesterday I flashed the new firmware version.

**Comment by S-Dafarra** [comment]
Here a log of the error that happened on one of the neck boards:
[log_icub-head_yarprobotinterface_2863_head_and_eyes.txt](https://github.com/robotology/icub-tech-support/files/7512151/log_icub-head_yarprobotinterface_2863_head_and_eyes.txt)

**Comment by marcoaccame** [comment]
```
SYS: the EOtheInfoDispatcher could not accept a  eOmn_info_properties_t item inside its transmitting queue. 
   > COMMENT by marco.accame: we lost 8 diagnostics messages because the tx FIFO was full
   > but what we have received is well enough.
SYS: the board is bootstrapping + . 
DEBUG: tag00 + RESTARTED after FATAL error 
DEBUG: tag00 + @ 3104150 ms 
DEBUG: tag00 + handler hw_BusFault, code 0x64 
DEBUG: tag00 + type see TBL 
DEBUG: tag00 + IRQHan BusFault Thread tmrma 
DEBUG: tag00 + ipsr 5, tid 3 
DEBUG: tag00 + MORE INFO 
DEBUG: tag00 + ICSR = 0x00400005 SHCSR = 0x00070082 
DEBUG: tag00 + CFSR = 0x00008200 HFSR = 0x00000000 
DEBUG: tag00 + DFSR = 0x00000000 MMFAR = 0x04e7fef3 
DEBUG: tag00 + BFAR = 0x04e7fef3 AFSR = 0x00000000 
DEBUG: tag00 + r0 = 0x080002e1 r1 = 0x04e7fee7 
DEBUG: tag00 + r2 = 0x20003eb0 r3 = 0xfee7fee7 
DEBUG: tag00 + r12 = 0x080002e5 lr = 0x08053945 
DEBUG: tag00 + pc = 0x08057cc0 psr = 0x2100020b 
DEBUG: tag00 + TM0 = 0x00000027 TM1 = 0x20007070 
DEBUG: tag00 + TM2 = 0x2000708c TM3 = 0x20007070 
DEBUG: tag00 + TM4 = 0x2000708c TM5 = 0x00000003 
DEBUG: tag00 + TM6 = 0x00000000 TM6 = 0x00000000 
DEBUG: tag00 + OS00 = 0x0000000a OS01 = 0x200073c0 
DEBUG: tag00 + OS02 = 0x20007388 OS03 = 0x00000004 
DEBUG: tag00 + OS04 = 0x00000000 OS05 = 0x00000064 
DEBUG: tag00 + OS06 = 0x000002d9 OS07 = 0x080002e1 
```



**Quite simply**
The values passed by high level function `osal_messagequeue_getquick()` come down inside the RTOS function and when copied into an automatic variable we can see that the **values have changed** 😮 💥. So, a valid pointer to an RTOS mailbox becomes a dirty pointer which writes out of memory and causes the fault.

**Brief explanation** 
At this point of the execution, the pointer `rtosobj` has value 0x20007388. We know it from ` OS02 = 0x20007388`.

```C++
    FATALERR_RT2_set(FT_0, 2);
    FATALERR_RT2_set(FT_2, rtosobj);    
    
    // any caller
    oosiit_mbx_retrieve(rtosobj, &p, s_osal_timeout2tick(tout));
```

Then execution goes through here, where `mailbox` is not changed. We also know that `message` is non NULL. Then we call the SVC handler with `__svc_oosiit_mbx_retrieve(mailbox, message, timeout)`.

```c++
extern oosiit_result_t oosiit_mbx_retrieve(oosiit_objptr_t mailbox, 
                                           void** message, uint32_t timeout)
{
    if(oosiit_res_NOK == s_oosiit_mbx_valid(mailbox))
    {
        return(oosiit_res_NOK);
    } 
    
    if(NULL == message)
    {
        return(oosiit_res_NOK);
    }
    
    if(0 != __get_IPSR()) 
    {   // inside isr
        FATALERR_RT2_set(FT_0, 31); // unlikely path
        return(isr_oosiit_mbx_retrieve(mailbox, message));
    } 
    else if(1 == s_oosiit_started)
    {   // call svc
         FATALERR_RT2_set(FT_0, 30); // detected path
         return(__svc_oosiit_mbx_retrieve(mailbox, message, timeout));
```
In here we are inside ethe SVC handler (we use a smaller stack also shared w/ the other IRQ handlers) and in here we call `iiitchanged_rt_mbx_wait(mailbox, message, timeout)`: 

```c++
extern oosiit_result_t svc_oosiit_mbx_retrieve(oosiit_objptr_t mailbox, 
                                               void** message, uint32_t timeout)
{
    OS_RESULT res;
    rt_iit_dbg_svc_enter();
    
    res = iitchanged_rt_mbx_wait(mailbox, message, timeout);
    
    rt_iit_dbg_svc_exit();
    return((oosiit_result_t)res);
}
```



And finally inside the crashing function `iitchanged_rt_mbx_wait()` the values have changed: `p_MCB` becomes  **0x00000004** and `message` becomes **0x00000000** (and we surely checked it to be not NULL!)

```c++
// in here we just use TIME_t for type of timeout and use iitchanged_rt_block() 
OS_RESULT iitchanged_rt_mbx_wait (OS_ID mailbox, void **message, TIME_t timeout) {
  /* Receive a message; possibly wait for it */
  P_MCB p_MCB = mailbox;
  P_TCB p_TCB;
    
    FATALERR_RT2_set(FT_0, 3);
    FATALERR_RT2_set(FT_3, p_MCB);
    FATALERR_RT2_set(FT_4, message);
```


**Why does it happen?**
The automatic variables use the stack, so I suspect it is a stack overflow. The RTOS functions run in [handler mode](https://developer.arm.com/documentation/ddi0439/b/Programmers-Model/Modes-of-operation-and-execution/Operating-modes)  because are run inside the [`SVCHandler()`](https://developer.arm.com/documentation/dui0203/j/handling-processor-exceptions/armv6-and-earlier--armv7-a-and-armv7-r-profiles/svc-handlers) , so they use the same stack as all the other interrupts. It may be that in some cases the stack runs out.

**What to do**
It is worth trying to slightly increase the stack assigned to the RTOS and rerun. The stack is now 8K. We may increase it to 11K, which is the maximum we can squeeze from RAM so far. I can also add further diagnostics to see the address of the automatic variables inside the RTOS so that if a crash still happens we can see if we go out of stack.

.

**Comment by marcoaccame** [comment]
Hi @S-Dafarra, a new version (v 3.41) for the binary of the `mc4plus` board is available for use on the robot. 

As usual it is named  `mc4plus-trace-rtos.hex` and is here: https://github.com/robotology/icub-firmware-build/tree/devel/ETH/MC4PLUS/bin/application. 

Pls use it and collect logs.

Thanks,

**Comment by S-Dafarra** [comment]
With @ale-git we flashed the FW yesterday. Today, one of the wrist boards had a fatal error even before the end of the calibration. Here the log:

[log_icub-head_yarprobotinterface_2805_left_wrist_dead.txt](https://github.com/robotology/icub-tech-support/files/7564662/log_icub-head_yarprobotinterface_2805_left_wrist_dead.txt)

**Comment by pattacini** [comment]
Just to make sure that we're aligned.
Did you also flash the latest MC4Plus FW?
Asking because for the incremental calibration only the EMS/2FOC FW was required.

**Comment by S-Dafarra** [comment]
> Just to make sure that we're aligned. Did you also flash the latest MC4Plus FW? Asking because for the incremental calibration only the EMS/2FOC FW was required and thus - maybe - you focused only on the latter boards.

Yes we flashed both the EMS and the MC4Plus

**Comment by pattacini** [comment]
Thanks!

cc @marcoaccame

**Comment by marcoaccame** [comment]
I have analyzed the new [log](https://github.com/robotology/icub-tech-support/issues/673#issuecomment-973075793) and ...
- the failure is exactly as the other times: a change in the values of arguments passed to a function by the RTOS
- I have verified (from the new log info) that we are not using stack memory below its limit, so it is not a stack underflow.

I omit the full analysis and I will focus only the new findings and on the actions.

**New findings**

So, I have looked for any possible cause of corruption in stack / argument values in the SVC calls. 

In some posts I have found that this may happen if the priorities of the SVC, SysTick and PostPend handler used by the RTOS are not correctly set. In such a case, if for instance the SVC is preempted then it could use a different stack and something could fail. That is mentioned for instance in [here](https://blog.stratifylabs.co/device/2013-10-12-Effective-Use-of-ARM-Cortex-M3-SVCall/), but also more importantly in an ARM's application note.

In https://developer.arm.com/documentation/ka003146/latest they say that that to **modify the NVIC priority grouping during runtime** may cause crashes.
> However, after that, the application fails to run correctly and may also crash after some time ......
> **CAUSE**
> One possible reason is the modification of the NVIC priority grouping during runtime. From the previous Keil RL-ARM RTX, you are probably aware, that such settings need to be done before the operating system is started e. g. in the main() function before the os_sys_init() call.

I remember that the wanted original design of the firmware is to perform the assignment of NVIC priority grouping only once at startup and before the start of the RTOS.
```C++
extern hal_result_t hal_sys_init(const hal_sys_cfg_t* cfg)
{
... omissis
    // set priority levels
    
    // configure once and only once the nvic to hold 4 bits for interrupt priorities and 0 for subpriorities
    // in stm32 lib ... NVIC_PriorityGroup_4 is 0x300, thus cmsis priority group number 3, thus
    // bits[7:4] for pre-emption priority and bits[3:0} for subpriority. but stm32 only has the 4 msb bits.
    // see page 114 of joseph yiu's book.
    NVIC_PriorityGroupConfig(NVIC_PriorityGroup_4);
    
    return(hal_res_OK);   
}
```
**Code Listing**. Initialization code of HAL called at startup which contains initialization of NVIC priorities (see https://github.com/robotology/icub-firmware/blob/master/emBODY/eBcode/arch-arm/libs/highlevel/abslayer/hal2/src/core/hal_sys.c#L151-L157).


But a double check on that has shown that there is a further change of NVIC priority grouping which is done in runtime after the RTOS initialization by the low level driver of the PWM. This is done only in the mc4plus and not in the ems and that may explain also why we never saw the fatal error in the ems.

```C++
extern hal_result_t hal_motor_init(hal_motor_t id, const hal_motor_cfg_t *cfg)
{
    
    if(hal_true == s_hal_motor_none_supported_is())
    {
        return(hal_res_NOK_generic);
    }
... omissis
	/* Configure one bit for preemption priority */
    NVIC_PriorityGroupConfig(NVIC_PriorityGroup_2);
... omissis
    return(hal_res_OK);   
}
```
**Code Listing**. Initialization code of the PWM driver which only the mc4plus calls at runtime at startup of the MC service (see  https://github.com/robotology/icub-firmware/blob/master/emBODY/eBcode/arch-arm/libs/highlevel/abslayer/hal2/src/extra/devices/hal_dc_motorctl.c#L361-L362).

**Actions**

The runtime change of the NVIC priorities is surely wrong and never intended, so it must be removed.  the application note on ARM's site suggests that it avoids possible crashes. I hope that it will solve our long running problem.

So, I will remove the runtime call of `NVIC_PriorityGroupConfig()`, test the new FW on a dedicated setup, release two PR on `icub-firmware` and `icub-firmware-build` and ask to use the new binaries

**Progress**
- [x] removal of  `NVIC_PriorityGroupConfig()`: done.
- [x] test of mc4plus binaries on a dedicated setup: done
- [ ] release of the two PRs: ongoing
- [ ] use of the new binaries: ongoing

**Comment by marcoaccame** [comment]
@S-Dafarra: merged. new binaries available.

**Comment by pattacini** [comment]
Hi @S-Dafarra 

Do you have any news to share on this?

**Comment by S-Dafarra** [comment]
Hi @pattacini! We flashed the firmware of https://github.com/robotology/icub-firmware-build/pull/42 about a week ago and we did not have any other issue since then. Fingers crossed :crossed_fingers:

**Comment by pattacini** [comment]
Hi @S-Dafarra 

So far so good on this?

**Comment by pattacini** [comment]
We need to clean up the debug messages we inserted to inspect the problem.

**Comment by pattacini** [comment]
Cleanup done in https://github.com/robotology/icub-firmware-build/pull/48.
Time to close!

**Comment by DanielePucci** [comment]
CC @gabrielenava

**Comment by marcoaccame** [comment]
> CC @gabrielenava

@DanielePucci ... is it ok, right? Or do we have a fatal error?

**Comment by gabrielenava** [comment]
>@DanielePucci ... is it ok, right? Or do we have a fatal error?

we don't have fatal errors at the moment were discussing this issue to understand if it was useful to update the firmware on iCubgenova04 MC4Plus
